### PR TITLE
BUG: Added missing error check in `ndarray.__contains__`

### LINKED ISSUE #15552: BUG: `array_contains` missing error check
Contains misses the error check for `any` failing:
```
    any = PyArray_Any((PyArrayObject *)res, NPY_MAXDIMS, NULL);
    Py_DECREF(res);
    ret = PyObject_IsTrue(any);
    Py_DECREF(any);
```
This leads to a segfault with objects that do not have a truthiness defined, as seen in pandas:
https://github.com/pandas-dev/pandas/issues/31922

### LINKED ISSUE #31922: BUG: segmentation fault with pd.NA and np.ndarray.__contains__
Testing for membership in an ndarray segfaults:

```
In [1]: pd.NA in np.array(["a"], dtype=object) 
Segmentation fault (core dumped)
```

Now, this might be something that needs to be fixed in numpy, as for plain lists it "works":

```
In [3]: pd.NA in ['a']  
...
TypeError: boolean value of NA is ambiguous
```

but numpy might not expect an error instead of True/False.

### PR DESCRIPTION
**Pull request by ganesh-k13**
Added NULL checks.

resolves #15552
resolves pandas-dev/pandas#31922

### CODE REVIEW

File: numpy/core/src/multiarray/sequence.c [hunk 1]
```diff
@@ -39,6 +39,11 @@ array_contains(PyArrayObject *self, PyObject *el)
         return -1;
     }
     any = PyArray_Any((PyArrayObject *)res, NPY_MAXDIMS, NULL);
+
+    if(any == NULL) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/sequence.c]:
```suggestion
    if (any == NULL) {
```

File: numpy/core/src/multiarray/sequence.c [hunk 2]
```diff
@@ -39,6 +39,11 @@ array_contains(PyArrayObject *self, PyObject *el)
         return -1;
     }
     any = PyArray_Any((PyArrayObject *)res, NPY_MAXDIMS, NULL);
+
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/sequence.c]:
```suggestion
```

**Review by eric-wieser — COMMENTED** [review]
Only style nits, otherwise looks good

File: numpy/core/src/multiarray/sequence.c (hunk 1 above)
Comment (ganesh-k13) [reply numpy/core/src/multiarray/sequence.c]: Thanks will do

**Comment by seberg** [comment]
Thanks, would you be happy to add a test as well? It will need to be an object which raises an error when you call `__bool__` and returns `self` for `__eq__` probably.

File: numpy/core/src/multiarray/sequence.c [hunk 3]
```diff
@@ -38,7 +38,12 @@ array_contains(PyArrayObject *self, PyObject *el)
     if (res == NULL) {
         return -1;
     }
+
     any = PyArray_Any((PyArrayObject *)res, NPY_MAXDIMS, NULL);
+    if (any == NULL) {
+        return -1;
+    }
+
     Py_DECREF(res);
```
Comment (seberg) [inline numpy/core/src/multiarray/sequence.c]: The res should be decref'd before the error check. It is not needed anymore in both branches.

**Comment by ganesh-k13** [comment]
Glad to add tests. I'm new here and did not quite understand when you say 
> raises an error when you call `__bool__`

File: numpy/core/src/multiarray/sequence.c (hunk 3 above)
Comment (seberg) [reply numpy/core/src/multiarray/sequence.c]: That is too far up :). You have to hold on to the object (read: the reference that gets decref'd) as long as you are using it. And it is used in the `PyArray_Any` call.

**Comment by seberg** [comment]
This is the implementation of a python class with the specific behaviour, such as:
```
class MyNA:
    def __bool__(self):
        raise ValueError("MyNA cannot be interpreted as a boolean value.")
```
Since the code will also call `MyNA() == MyNA()` here, you also have to implement `__eq__` check the special dunder methods when creating a python class.

File: numpy/core/src/multiarray/sequence.c (hunk 3 above)
Comment (ganesh-k13) [reply numpy/core/src/multiarray/sequence.c]: Ohh sorry, my bad, will fix it. Thanks.

**Comment by ganesh-k13** [comment]
Guys, I can't see a workaround, it's working in the shell but failing in the test script:

```
class CustomNA:
	def __bool__(self):
		raise TypeError("CustomNA cannot be interpreted as a boolean value.")

	def __eq__(self, other):
		return self
CustomNA() in np.array(["a"], dtype=object)
```
>  DeprecationWarning: elementwise comparison failed; this will raise an error in the future.

In shell:
```
>>> CustomNA() not in np.array(["a"], dtype=object)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 5, in __bool__
TypeError: CustomNA cannot be interpreted as a boolean value.
```

Can you please help me. Apologies if it's a very basic doubt.

**Comment by seberg** [comment]
No, that is not basic at all... Maybe Eric already anticipated this. The whole issue is far more complicated than it appears, and the DeprecationWarning (which is raised as an error in the test, you will see identical results with `import warnings; warnings.simplefilter("error")`) makes things tricky.

The "correct" test could be hacked by adding `__array_priority__ = 1` (which also shows how badly broken `__contains__` is in yet another way... 

Maybe Eric anticipated this. I am OK with just putting it in without a test also, it gets complicated to cover the exact code path you are hitting... But you can add it referencing this issue with `gh-15553` in a comment so that there is at least some fighting chance of understanding what that test does/did...

**Comment by ganesh-k13** [comment]
Oh I see, I guess we can add that UT after fixing the main issue.

**Comment by seberg** [comment]
OK, I will just merge this for now. If anyone feels we should add the test with `__array_priority__` just comment and I will do it, but it does feel a bit complicated.

The test failures are unrelated due to some gfortran issue.

**Comment by ganesh-k13** [comment]
Thanks for the merge @seberg, I'll try it myself and ask here if I'm stuck.

**Pull request merged**